azurelinuxagent/ga/firewall_manager.py

```diff
@@ -0,0 +1,499 @@
+
+    def _get_drop_rule_command(self, command_option):
+        return self._base_command + [command_option, "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "conntrack", "--ctstate", "INVALID,NEW", "-j", "DROP"]
+
+    def _get_legacy_rule_command(self, command_option):
+        # There was a rule change at 2.2.26, which started dropping non-root traffic to WireServer. The previous rule allowed traffic, and needs to be removed
+        # for the newer DROP rule to have any effect. This function returns the command to manipulate the legacy rule that was added <= 2.2.25. Until 2.2.25
+        # has aged out, keep this cleanup in place.
+        return self._base_command + [command_option, "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "conntrack", "--ctstate", "INVALID,NEW", "-j", "ACCEPT"]
+
+    def _get_append_command_option(self):
+        return "-A"
+
+    def _get_check_command_option(self):
+        return "-C"
+
+    def _get_delete_command_option(self):
+        return "-D"
+
+
+class FirewallCmd(_FirewallManagerMultipleRules):
+    """
+    FirewallManager based on the firewalld command-line tool.
+    """
+    def _get_state_command(self):
+        return ["firewall-cmd", "--permanent", "--direct", "--get-all-passthroughs"]
+
+    def _get_accept_dns_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", '--destination-port', '53', '-j', 'ACCEPT']
+
+    def _get_accept_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "owner", "--uid-owner", str(os.getuid()), "-j", "ACCEPT"]
+
+    def _get_drop_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "conntrack", "--ctstate", "INVALID,NEW", "-j", "DROP"]
+
+    def _get_legacy_rule_command(self, command_option):
+        # Agents <= 2.7.0.6 inserted (-I) the rule to accept DNS traffic; later agents changed that to append (-A) the rule.
+        # The insert rule needs to be removed, otherwise there will be duplicate rules for DNS.
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-I", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", '--destination-port', '53', '-j', 'ACCEPT']
+
+    def _get_append_command_option(self):
+        return "--passthrough"
+
+    def _get_check_command_option(self):
+        return "--query-passthrough"
+
+    def _get_delete_command_option(self):
+        return "--remove-passthrough"
+
+
+class NfTables(FirewallManager):
```

Comment:
FirewallManager.create() instantiate the correct class depending on the tools present on the system. The derived classes can be instantiated directly when the caller already knows what tool to use.

Reply:
Not quite. Current code tries iptables, then nft, then raises if none present

```
        try:
            manager = IpTables(wire_server_address)
            logger.info("Using iptables to manage firewall rules")
            return manager
        except FirewallManagerNotAvailableError:
            pass

        try:
            manager = NfTables(wire_server_address)
            logger.info("Using nftables to manage firewall rules")
            return manager
        except FirewallManagerNotAvailableError:
            pass

        raise FirewallManagerNotAvailableError("Cannot create a firewall manager; no known command-line tools are available")

```

Reply:
Assume iptables and nftables both tools not present. First, we try iptables and got FirewallManagerNotAvailableError. Then we try nft but who is throwing  FirewallManagerNotAvailableError error when we initialize NfTables(wire_server_address). It just takes nftables as firewall manger and try to setup the rules, will get different exception during setup.

Reply:
Ah, I see what you mean. NfTables is missing its init method to do the check. I'll add it. Thanks.

Reply:
Thanks again for pointing this out. 

Both the firewall-cmd and nft implementations were not honoring the contract required by the base class:

`    If a concrete implementation cannot be instantiated because the underlying command-line tool is not available, it must raise a FirewallManagerNotAvailable exception.
` 

This is not an issue in the currently supported distros (and the end-to-end tests prove that). but it could be an issue in the future, and unit tests should have caught this, except there were no unit tests for this!

Issue has been fixed now, and unit tests added.